### PR TITLE
Add Euler's Totient Function Implementation

### LINKED ISSUE #63: Implement Euler's Totient Function
### Challenge

The goal of this challenge is to implement Euler's Totient function (ϕ(n)), which counts how many integers between 1 and n (inclusive) are coprime with n. Two numbers are considered coprime if their greatest common divisor (GCD) equals 1.

Your task is to write the function euler_totient(n) that calculates the Euler's Totient function for any positive integer n.

### Steps to Achieve This

Understand the Euler's Totient Function
The Totient function, also known as the phi function, returns the count of numbers from 1 to n that are coprime with n.
The formula is derived by considering the prime factorization of n and adjusting the result based on its prime factors.
### Define the Function
Create a function euler_totient(n) that takes an integer n as input and returns the Totient value ϕ(n).
Initialize the result with n.
Loop through integers starting from 2 up to the square root of n and check if n is divisible by these integers.
If n is divisible by a prime factor, reduce n by dividing it by the factor, and adjust the result by subtracting the appropriate value for that factor.
If there’s any prime factor greater than √n left, adjust the result accordingly.
### Edge Cases
Handle the special case when n = 1, since ϕ(1) = 1.
Ensure that your function works for prime numbers, composite numbers, and squares of primes.
If the input n is zero or negative, raise a ValueError.

### Testing
Write unit tests that cover the following cases:
Basic examples like ϕ(9), ϕ(10), and ϕ(1).
Test for prime numbers and composite numbers.
Ensure your function handles edge cases such as n = 1, n = 0, and negative values.
Test that the function raises the appropriate errors for invalid inputs.
Example Usage:

>>> euler_totient(9)
6

>>> euler_totient(10)
4

>>> euler_totient(1)
1
Additional Notes:

When n is prime, the result is n - 1 because all numbers less than n are coprime with n.
When n is a composite number, the function should properly calculate the result based on its prime factors.

### Expected Outcome
After completing this challenge, the function should correctly compute the Totient function for any valid input and return the expected results.

### PR DESCRIPTION
**Pull request by RandomProjects-db**
The Euler's Totient function (ϕ(n)) counts the number of integers between 1 and n that are coprime with n. Two numbers are considered coprime if their greatest common divisor (GCD) equals 1. The function has applications in number theory and cryptography.

This pull request adds the implementation of the Euler Totient function, including docstrings with examples and unit tests. The euler_totient function computes the value of Euler's Totient function for a given integer n, and the implementation runs in O(√n) time complexity.

### Changes

Added a function euler_totient(n) to calculate Euler's Totient function.
Included docstrings with example usage.
Created unit tests to validate the implementation for various cases, including edge cases and invalid inputs.

### Test Cases Include
Prime numbers.
Composite numbers.
The case for n = 1.
Edge cases for invalid input values such as negative numbers and zero.

###cExample Usage

>>> euler_totient(9)
6

>>> euler_totient(10)
4

### Testing
Unit tests are added in test_euler_totient.py and tested using unittest.

closes #63

### CODE REVIEW

File: solutions/euler_totient.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    >>> euler_totient(12)
+    4
+
+    >>> euler_totient(1)
+    1
+
+    >>> euler_totient(5)
+    4
+    """
```
Comment (Bikaze) [inline solutions/euler_totient.py:41]: Please add a doctest for an assertion error for comprehensiveness.

File: solutions/euler_totient.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    # Defensive assertions
+    if not isinstance(n, int):
+        raise ValueError("Input must be an integer.")
+    if n <= 0:
+        raise ValueError("Input must be a positive integer.")
```
Comment (Bikaze) [inline solutions/euler_totient.py:46]: It is better to use an assert statement here rather than an if statement.

File: solutions/euler_totient.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    >>> euler_totient(1)
+    1
+
+    >>> euler_totient(5)
+    4
+    """
+    # Defensive assertions
+    if not isinstance(n, int):
+        raise ValueError("Input must be an integer.")
```
Comment (Bikaze) [inline solutions/euler_totient.py:44]:
```suggestion
    assert isinstance(n, int), "Input must be an integer."
```

File: solutions/tests/test_euler_totient.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    It tests the function's ability to handle various valid inputs, including edge cases like 1,
+    and invalid inputs such as non-integer values, negative numbers, and zero.
+    """
+
+    def test_valid_input(self):
+        """Test case for valid positive integer input."""
+        self.assertEqual(euler_totient(9), 6)
+        self.assertEqual(euler_totient(12), 4)
+        self.assertEqual(euler_totient(5), 4)
+        self.assertEqual(euler_totient(1), 1)
```
Comment (Bikaze) [inline solutions/tests/test_euler_totient.py:35]: Each test function should have one assertion.

File: solutions/tests/test_euler_totient.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        """Test case for zero input."""
+        with self.assertRaises(ValueError):
+            euler_totient(0)
+
+    def test_invalid_input_non_integer(self):
+        """Test case for non-integer input."""
+        with self.assertRaises(ValueError):
+            euler_totient("string")
+        with self.assertRaises(ValueError):
+            euler_totient(3.5)
```
Comment (Bikaze) [inline solutions/tests/test_euler_totient.py:52]: Each test function should have one assertion.

File: solutions/tests/test_euler_totient.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_large_input(self):
+        """Test case for a large positive integer."""
+        self.assertEqual(euler_totient(1000), 400)
+
+    def test_prime_number(self):
+        """Test case for prime number input."""
+        self.assertEqual(euler_totient(7), 6)
+        self.assertEqual(euler_totient(13), 12)
```
Comment (Bikaze) [inline solutions/tests/test_euler_totient.py:61]: Each test function should have one assertion.

**Review by Bikaze — COMMENTED** [review]
Wow this is a good challenge! Fix the issues I highlighted as per the code review requirements and you should be good to go.

File: solutions/euler_totient.py (hunk 1 above)
Comment (RandomProjects-db) [reply solutions/euler_totient.py:44]: fixed!

File: solutions/euler_totient.py (hunk 3 above)
Comment (RandomProjects-db) [reply solutions/euler_totient.py]: fixed with:     
assert isinstance(n, int), "Input must be an integer."
assert n > 0, "Input must be a positive integer."

File: solutions/tests/test_euler_totient.py (hunk 5 above)
Comment (RandomProjects-db) [reply solutions/tests/test_euler_totient.py]: fixed

File: solutions/tests/test_euler_totient.py (hunk 6 above)
Comment (RandomProjects-db) [reply solutions/tests/test_euler_totient.py]: fixed

File: solutions/euler_totient.py (hunk 2 above)
Comment (RandomProjects-db) [reply solutions/euler_totient.py]: assert statements are typically used for debugging and are often disabled when the program is run in optimized mode (using the -O flag), which would skip those checks entirely.
Instead, raising a ValueError is more appropriate for input validation, as it will always be checked regardless of optimization settings. Additionally, the assertion redundancy (first checking the type and then asserting the same condition) can be avoided.

or it is required by the code review requirements ?

**Review by Bikaze — APPROVED** [review]
Wooohooo! You're kill'n it.

File: solutions/euler_totient.py (hunk 2 above)
Comment (Bikaze) [reply solutions/euler_totient.py]: Yes you are right. I leave for to you to decide.

Comment (Bikaze) [reply solutions/euler_totient.py]: resolve the conversation for the merge to take place.
